Fallback to customized Product Archive template if no better Category or Tag templates are present

sunyatasattva opened the pull request:
If the user has customized the `archive-product` template and there is no better choice for the Category
or Tag, then the custom Product Archive is chosen.

The priority order is the following:

1. Custom Taxonomy Template
2. Taxonomy Template from the Theme
3. Custom Product Archive Template
4. Product Archive Template from the Theme

Fixes woocommerce/woocommerce#42656


P.S. Not sure about the labels here.

### Screenshots

<img width="1030" alt="Screen Shot 2022-01-13 at 23 38 57" src="https://user-images.githubusercontent.com/1847066/149420194-586ada11-399f-4c5c-923f-e2a5bf9f7d70.png">

### Testing

#### Manual Testing

How to test the changes in this Pull Request:

1. Enable a theme with an `archive-product.html` but no taxonomy templates (or add your own to the theme).
2. Go to Appearance → Editor → Browse All Templates.
3. Notice that the Category and Tag templates are also available from the theme (previous behaviour).
4. Customize `archive-product` and save.
5. Notice that the Category and Tag templates now also appear as customized and are applied the same edits.

> **Note:** Even though they will have the customized blue dot next to them signaling to the user that some edit has been implemented, they will not appear at the top of the list because they have not received any direct customization. Because of the priority order above (template from theme takes prio over custom product archive), doing otherwise would make things messy. However, happy to talk about this in detail if required.

### User Facing Testing
These are steps for user testing (where "user" is someone interacting with this change that is not editing any code).
* [x] Same as above


### Changelog

> Customizations to the theme Product Archive block template are applied to Category and Tag templates is no better alternative is available

sunyatasattva committed:
Fallback to customized Product Archive template if no better Category or Tag templates are present

If the user has customized the `archive-product` template and there is no better choice for the Category
or Tag, then the custom Product Archive is chosen.

The priority order is the following:

1. Custom Taxonomy Template
2. Taxonomy Template from the Theme
3. Custom Product Archive Template
4. Product Archive Template from the Theme

Fixes #5520

tjcafferkey commented:
Few things to note:
* If I enable a theme that does **not** have the `archive-product.html` template and gets them provided from Woo Blocks. The templates screen is missing Product Category and Product Tag templates.
* Customizations of the themes `archive-product.html` do not render on the frontend for product category/tag pages.
* Customizations of the themes `archive-product.html` template do not appear in the Site Editor for product category/tag templates.

Not sure if I'm doing anything wrong here?

tjcafferkey commented on src/BlockTemplatesController.php:
```diff
@@ -341,10 +341,22 @@ function ( $template ) use ( $template_slug ) {
 			if ( BlockTemplateUtils::template_is_eligible_for_product_archive_fallback( $template_slug ) ) {
-				$template_file = BlockTemplateUtils::get_theme_template_path( 'archive-product' );
-				$templates[]   = BlockTemplateUtils::create_new_block_template_object( $template_file, $template_type, $template_slug, true );
+				$customized_archive_template_idx = array_search( 'archive-product', array_column( $already_found_templates, 'slug' ), true );
```
Can we rename `idx` to `index` please?

sunyatasattva committed:
Rename idx to index

sunyatasattva committed:
Make sure the fallback template is rendered in the editor correctly

sunyatasattva committed:
Make sure the fallback template is rendered in the front-end correctly

sunyatasattva committed:
Make sure to fallback to templates provided by us in the worse case scenario

sunyatasattva commented:
@tjcafferkey I think I addressed all the concerns and I have tested more extensively. Now all the three cases you pointed out should work. Thanks for noticing them 🤦‍♀️

tjcafferkey commented:
When I created a `archive-product.html` file in my theme, and then customized it. No products were rendering on the frontend for neither the shop archive, or tags/categories.

sunyatasattva commented:
@tjcafferkey Could you elaborate on what you do exactly? Are you sure that in your customizations you have the legacy block activated?

I tried to reproduce in several ways, even completely deactivating and cloning from scratch. Here is a video of the flow working for me:

https://user-images.githubusercontent.com/1847066/150039301-89f931e3-51f7-4fdf-bde5-71f7b00bf085.mov

Perhaps I'm getting something wrong?

sunyatasattva committed:
Fix strip file suffix function name

tjcafferkey commented:
>Perhaps I'm getting something wrong?

I've tried to replicate again but can't so it looks like it will have been an issue on my end. 

However I've done some more testing and there a number of bugs which I've found, please see my recording. Happy to go through this with you on a call if needed, looks like this particular issue might be a little more complex than we first thought.

https://user-images.githubusercontent.com/8639742/150152612-8cfca6fb-e179-4049-b9ba-e75efe33ca03.mp4

sunyatasattva commented:
1:48 I was like: 😱 Whaat?

Thanks for the extensive testing, @tjcafferkey. I'm going to be back to the drawing board.

sunyatasattva commented:
We had postponed looking into this and its complex behaviours because we decided to go on with E2E tests and make sure such random edge cases could not happen as we focus on this. Perhaps it's time for me to pick this up again :)

sunyatasattva committed:
Remove `taxonomy-product_cat.html` from the mock theme

This is because we want to use the mock theme as
a testbed for template override logic.

sunyatasattva committed:
Add tests for Store Editing Template Fallback logic
